### PR TITLE
Fix 943

### LINKED ISSUE #943: Convert method for StaticArray shouldn't access data field
https://github.com/JuliaArrays/StaticArrays.jl/pull/550 introduces https://github.com/JuliaArrays/StaticArrays.jl/blob/1d5524cc131f8ff9014f4625d68c72a48e35a9df/src/convert.jl#L6 that accesses `data` for `StaticArray` but since the latter is an abstract type, there is no reason to believe that it has a `data` field. This is e.g. invalid for https://github.com/SciML/LabelledArrays.jl/blob/bd9cd38dc9f4ef4eb366c16492fa64ca795eb566/src/slarray.jl#L1-L19

### PR DESCRIPTION
**Pull request by mateuszbaran**
This fixes #943 . It also changes the copy constructor for `SizedArray` so that it actually makes a copy, just like normal `Array`s.

### CODE REVIEW

**Comment by andyferris** [comment]
I don’t particularly mind, but it was intended to behave like a `SubArray`, and just present a “view” with static indices without copying the data.

**Comment by mateuszbaran** [comment]
I can change this but it's a bit hard to come up with a consistent set of rules that don't deviate too far from the current behavior. `SizedArray` doesn't currently behave like a view for `MArray` or `SArray` for example.

**Review by hyrodium — APPROVED** [review]
LTGM

**Comment by hyrodium** [comment]
What's the status of this PR? Is this just waiting to resolve the conflicts?

**Comment by mateuszbaran** [comment]
The status is that I'm a bit worried that some people using `SizedArray` may complain about the change regarding copy constructor of `SizedArray`. And the original issue is easy to fix downstream so it's not a particularly big issue.

**Comment by hyrodium** [comment]
Is there any simple example for the issues around the copy constructor of `SizedArray`?
The following `m0`, `m1` and `m2` share the same memory, both in current `master` and `mbaran/fix943`.
```julia
julia> using StaticArrays

julia> m0 = rand(3,3)
3×3 Matrix{Float64}:
 0.66047   0.805579  0.72429
 0.72589   0.950237  0.00648303
 0.317448  0.70089   0.779554

julia> m1 = SizedMatrix{3,3}(m0)
3×3 SizedMatrix{3, 3, Float64, 2, Matrix{Float64}} with indices SOneTo(3)×SOneTo(3):
 0.66047   0.805579  0.72429
 0.72589   0.950237  0.00648303
 0.317448  0.70089   0.779554

julia> m2 = SizedMatrix{3,3}(m1)
3×3 SizedMatrix{3, 3, Float64, 2, Matrix{Float64}} with indices SOneTo(3)×SOneTo(3):
 0.66047   0.805579  0.72429
 0.72589   0.950237  0.00648303
 0.317448  0.70089   0.779554

julia> m0[1] = 5
5

julia> m1
3×3 SizedMatrix{3, 3, Float64, 2, Matrix{Float64}} with indices SOneTo(3)×SOneTo(3):
 5.0       0.805579  0.72429
 0.72589   0.950237  0.00648303
 0.317448  0.70089   0.779554

julia> m2
3×3 SizedMatrix{3, 3, Float64, 2, Matrix{Float64}} with indices SOneTo(3)×SOneTo(3):
 5.0       0.805579  0.72429
 0.72589   0.950237  0.00648303
 0.317448  0.70089   0.779554
```
Sorry if the above example is not appropriate.

**Comment by mateuszbaran** [comment]
I've updated locally the PR to the new master and it works differently:
```julia
julia> using StaticArrays
[ Info: Precompiling StaticArrays [90137ffa-7385-5640-81b9-e52037218182]

julia> m0 = rand(3,3)
3×3 Matrix{Float64}:
 0.80775    0.752221  0.358652
 0.752796   0.582847  0.842556
 0.0938127  0.991837  0.664063

julia> m1 = SizedMatrix{3,3}(m0)
3×3 SizedMatrix{3, 3, Float64, 2, Matrix{Float64}} with indices SOneTo(3)×SOneTo(3):
 0.80775    0.752221  0.358652
 0.752796   0.582847  0.842556
 0.0938127  0.991837  0.664063

julia> m2 = SizedMatrix{3,3}(m1)
3×3 SizedMatrix{3, 3, Float64, 2, Matrix{Float64}} with indices SOneTo(3)×SOneTo(3):
 0.80775    0.752221  0.358652
 0.752796   0.582847  0.842556
 0.0938127  0.991837  0.664063

julia> m1.data === m2.data
false

julia> m0[1] = 12
12

julia> m1
3×3 SizedMatrix{3, 3, Float64, 2, Matrix{Float64}} with indices SOneTo(3)×SOneTo(3):
 12.0        0.752221  0.358652
  0.752796   0.582847  0.842556
  0.0938127  0.991837  0.664063

julia> m2
3×3 SizedMatrix{3, 3, Float64, 2, Matrix{Float64}} with indices SOneTo(3)×SOneTo(3):
 0.80775    0.752221  0.358652
 0.752796   0.582847  0.842556
 0.0938127  0.991837  0.664063
```

**Comment by mateuszbaran** [comment]
Just in case I've uploaded my update but it doesn't even pass tests now.

File: src/SizedArray.jl [hunk 1]
```diff
@@ -15,11 +15,11 @@ end
 @inline (::Type{SZA})(a::AbstractArray) where {SZA<:SizedArray} = construct_type(SZA, a)(a)
 
 # disambiguation
-@inline SizedArray{S,T,N,M}(a::StaticArray) where {S<:Tuple,T,N,M} = construct_type(SizedArray{S,T,N,M}, a)(a.data)
-@inline SizedArray{S,T,N}(a::StaticArray) where {S<:Tuple,T,N} = construct_type(SizedArray{S,T,N}, a)(a.data)
-@inline (::Type{SZA})(a::StaticArray) where {SZA<:SizedArray} = construct_type(SZA, a)(a.data)
+@inline SizedArray{S,T,N,M}(a::StaticArray) where {S<:Tuple,T,N,M} = construct_type(SizedArray{S,T,N,M}, a)(Tuple(a))
+@inline SizedArray{S,T,N}(a::StaticArray) where {S<:Tuple,T,N} = construct_type(SizedArray{S,T,N}, a)(Tuple(a))
+@inline (::Type{SZA})(a::StaticArray) where {SZA<:SizedArray} = construct_type(SZA, a)(Tuple(a))
```
Comment (hyrodium) [inline src/SizedArray.jl:20]:
```suggestion
@inline SizedArray{S,T,N,M}(a::StaticArray) where {S<:Tuple,T,N,M} = construct_type(SizedArray{S,T,N,M}, a)(Tuple(a))
@inline SizedArray{S,T,N}(a::StaticArray) where {S<:Tuple,T,N} = construct_type(SizedArray{S,T,N}, a)(Tuple(a))
@inline (::Type{SZA})(a::StaticArray) where {SZA<:SizedArray} = construct_type(SZA, a)(Tuple(a))
@inline SizedArray{S,T,N,M}(a::SizedArray) where {S<:Tuple,T,N,M} = construct_type(SizedArray{S,T,N,M}, a)(a.data)
@inline SizedArray{S,T,N}(a::SizedArray) where {S<:Tuple,T,N} = construct_type(SizedArray{S,T,N}, a)(a.data)
@inline (::Type{SZA})(a::SizedArray) where {SZA<:SizedArray} = construct_type(SZA, a)(a.data)
```
How about adding methods like this?

File: src/SizedArray.jl [hunk 2]
```diff
@@ -15,11 +15,11 @@ end
...
 # TODO: Should we respect `TData`?
-SizedArray{S,T,N,M,TData}(a::TData) where {S<:Tuple,T,N,M,TData<:StaticArray{<:Tuple,T,M}} = SizedArray{S,T,N,M}(a.data)
+SizedArray{S,T,N,M,TData}(a::TData) where {S<:Tuple,T,N,M,TData<:StaticArray{<:Tuple,T,M}} = SizedArray{S,T,N,M}(Tuple(a))
```
Comment (hyrodium) [inline src/SizedArray.jl:22]:
```suggestion
SizedArray{S,T,N,M,TData}(a::TData) where {S<:Tuple,T,N,M,TData<:StaticArray{<:Tuple,T,M}} = SizedArray{S,T,N,M}(Tuple(a))
SizedArray{S,T,N,M,TData}(a::TData) where {S<:Tuple,T,N,M,TData<:SizedArray{<:Tuple,T,M}} = SizedArray{S,T,N,M}(a.data)
```

**Comment by andyferris** [comment]
It clearly should use `Tuple` in the general `StaticArray` case, so we should merge some variant of this. I don't mind either way about copying of data; ideally users would write `SizedArray{...}(other_sized_array.data)` if they wanted to ensure aliasing!

If people in the wild _really_ do rely on the current behavior, we can just make a major release with a note about this rather than make our code more complex.

**Comment by mateuszbaran** [comment]
Sorry for a delay, I will try to go back to this as soon as possible.

File: src/SizedArray.jl (hunk 1 above)
Comment (mateuszbaran) [reply src/SizedArray.jl:20]: Yes, this might work.

**Comment by mateuszbaran** [comment]
I've taken a look at missing code coverage, and then at the code again. The current version doesn't seem right. Feel free to continue working on it but I'm not going to merge something I'm not happy with, and I may never be happy with any variant of this.

**Pull request closed without merging**